File: java/3. Longest Substring Without Repeating Characters/Solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> seen = new HashSet<>();

        int start = 0;
        int current = 0;
        int maxLength = 0;

        while (current < s.length()) {
            char currentChar = s.charAt(current);
            if (seen.contains(currentChar)) {
                seen.remove(s.charAt(start++));
                continue;
            }

            seen.add(currentChar);
            maxLength = Math.max(maxLength, current - start + 1);
            current++;
        }

        return maxLength;
    }
}

File: python/3. Longest Substring Without Repeating Characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        seen = set()

        start = 0
        current = 0
        maxLength = 0

        while current < len(s):
            currentChar = s[current]
            if currentChar in seen:
                seen.remove(s[start])
                start += 1
                continue

            seen.add(currentChar)
            maxLength = max(maxLength, current - start + 1)
            current += 1

        return maxLength


